queue/queue.mbt

```diff
@@ -0,0 +1,364 @@
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+/// ```
+pub fn Queue::from_array[T](array : Array[T]) -> Queue[T] {
+  let queue : Queue[T] = new()
+  for i = 0; i < array.length(); i = i + 1 {
+    queue.add(array[i])
+  }
+  queue
+}
+
+test "from_array" {
+  let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+  @assertion.assert_eq(queue.length, 4)?
+  @assertion.assert_eq(queue.pop(), 1)?
+  @assertion.assert_eq(queue.pop(), 2)?
+  @assertion.assert_eq(queue.pop(), 3)?
+  @assertion.assert_eq(queue.pop(), 4)?
+  @assertion.assert_eq(queue.length, 0)?
+}
+
+/// Creates a new queue from a list.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::from_list(Cons(1, Cons(2, Cons(3, Cons(4, Nil))))
+/// ```
+pub fn Queue::from_list[T](list : List[T]) -> Queue[T] {
+  let queue : Queue[T] = new()
+  loop list {
+    l =>
+      match l {
+        Cons(head, tail) => {
+          queue.add(head)
+          continue tail
+        }
+        Nil => break
+      }
+  }
+  queue
+}
+
+test "from_list" {
+  let queue : Queue[Int] = Queue::from_list(
+    Cons(1, Cons(2, Cons(3, Cons(4, Nil)))),
+  )
+  @assertion.assert_eq(queue.length, 4)?
+  @assertion.assert_eq(queue.pop(), 1)?
+  @assertion.assert_eq(queue.pop(), 2)?
+  @assertion.assert_eq(queue.pop(), 3)?
+  @assertion.assert_eq(queue.pop(), 4)?
+  @assertion.assert_eq(queue.length, 0)?
+}
+
+/// Clears the queue.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+/// queue.clear()
+/// ```
+pub fn clear[T](self : Queue[T]) -> Unit {
+  self.length = 0
+  self.tail = None
+}
+
+test "clear" {
+  let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+  queue.clear()
+  @assertion.assert_eq(queue.length, 0)?
+  @assertion.assert_true(queue.tail == None)?
+}
+
+/// Adds a value to the queue.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::create()
+/// queue.add(1)
+/// ```
+pub fn add[T](self : Queue[T], value : T) -> Unit {
+  if self.length == 0 {
+    let cell : Cell[T] = { content: value, next: None }
+    cell.next = Some(cell)
+    self.length = 1
+    self.tail = Some(cell)
+  } else {
+    let tail = self.tail.unwrap()
+    let head = tail.next
+    let cell = { content: value, next: head }
+    self.length = self.length + 1
+    tail.next = Some(cell)
+    self.tail = Some(cell)
+  }
+}
+
+test "add" {
+  let queue : Queue[Int] = new()
+  queue.add(1)
+  queue.add(2)
+  queue.add(3)
+  queue.add(4)
+  @assertion.assert_eq(queue.length, 4)?
+  @assertion.assert_eq(queue.pop(), 1)?
+  @assertion.assert_eq(queue.pop(), 2)?
+  @assertion.assert_eq(queue.pop(), 3)?
+  @assertion.assert_eq(queue.pop(), 4)?
+  @assertion.assert_eq(queue.length, 0)?
+}
+
+/// Peeks at the first value in the queue.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+/// let first = queue.peek()
+/// ```
+pub fn peek[T](self : Queue[T]) -> Option[T] {
+  let tail = self.tail
+  match tail {
+    Some(tail) => {
```

Comment:
1. we support nested pattern match
2. this seems to be incorrect?
`Some ({content}) ->`